File: sorts/insertion_sort_list.py
class Solution:
    def insertionSortList(self, head: Optional[ListNode]) -> Optional[ListNode]:
      if not head or not head.next:
        return head

    dummy = ListNode(0)
    curr = head

    while curr:
      next_node = curr.next
      
      prev = dummy
      while prev.next and prev.next.val < curr.val
        prev = prev.next


      curr.next = prev.next
      prev.next = curr

      curr = next_node


    return dummy.next

File: sorts/insertion_sort_list.rb
class ListNode
    attr_accessor :val, :next
    def initialize(val = 0, next_node = nil)
        @val = val
        @next = next_node
    end
end

def insertion_sort_list(head)
  return head if head.nil? || head.next.nil?

  dummy = ListNode.new(0)
  curr = head

  while curr
    next_node = curr.next
    
    prev = dummy
    while prev.next && prev.next.val < curr.val
      prev = prev.next
    end

    curr.next = prev.next
    prev.next = curr

    curr = next_node
  end
  
  dummy.next
end



# def create_linked_list(arr)
#     return nil if arr.empty?
    
#     head = ListNode.new(arr[0])
#     current = head
    
#     (1...arr.size).each do |i|
#         current.next = ListNode.new(arr[i])
#         current = current.next
#     end
    
#     head
# end



# insertion_sort_list(create_linked_list([4,2,1,3]))

